Source/WebCore/platform/mediastream/mac/ScreenCaptureKitSharingSessionManager.mm

```diff
@@ -242,82 +279,298 @@ static bool screenCaptureKitPickerFeatureEnabled()
+#if HAVE(SC_CONTENT_SHARING_PICKER)
+    else
+        stream = adoptNS([PAL::allocSCStreamInstance() initWithSharingSession:m_pendingSession.get() captureOutputProperties:configuration delegate:(id<SCStreamDelegate> _Nullable)delegate]);
+#endif
+
+    if (!stream) {
+        RELEASE_LOG_ERROR(WebRTC, "ScreenCaptureKitSharingSessionManager::createSessionSourceForDevice - unable to create SCStream.");
+        return nullptr;
+    }
+
+    auto cleanupFunction = [this, weakThis = WeakPtr { *this }](ScreenCaptureSessionSource& source) mutable {
+        if (!weakThis)
+            return;
+
+        cleanupSessionSource(source);
+    };
+
+    auto newSession = ScreenCaptureSessionSource::create(WTFMove(observer), WTFMove(stream), WTFMove(m_pendingContentFilter), WTFMove(m_pendingSession), WTFMove(cleanupFunction));
+    m_activeSources.append(newSession);
+
+    return newSession;
+}
+
+void ScreenCaptureKitSharingSessionManager::cleanupSessionSource(ScreenCaptureSessionSource& source)
+{
+    auto index = m_activeSources.findIf([&source](auto activeSource) {
+        return &source == activeSource.get();
     });
-    ASSERT(index != notFound);
     if (index == notFound) {
-        RELEASE_LOG_ERROR(WebRTC, "ScreenCaptureKitSharingSessionManager::takeSharingSessionForFilter sharing session not found");
-        return nullptr;
+        RELEASE_LOG_ERROR(WebRTC, "ScreenCaptureKitSharingSessionManager::cleanupSessionSource - unknown source.");
+        return;
     }
 
-    RetainPtr<SCContentSharingSession> session = m_pendingCaptureSessions[index];
-    m_pendingCaptureSessions.remove(index);
-    [m_promptHelper stopObservingSession:session.get()];
+    auto activeSources = m_activeSources[index];
+    m_activeSources.remove(index);
 
-    return session;
+    if (!activeSources)
+        return;
+
+    auto sharingSession = activeSources->sharingSession();
+    if (!sharingSession)
+        return;
+
+    if (m_promptHelper)
+        [m_promptHelper stopObservingSession:sharingSession.get()];
+    [sharingSession end];
 }
 
+Ref<ScreenCaptureSessionSource> ScreenCaptureSessionSource::create(WeakPtr<Observer> observer, RetainPtr<SCStream> stream, RetainPtr<SCContentFilter> filter, RetainPtr<SCContentSharingSession> sharingSession, CleanupFunction&& cleanupFunction)
+{
+    return adoptRef(*new ScreenCaptureSessionSource(WTFMove(observer), WTFMove(stream), WTFMove(filter), WTFMove(sharingSession), WTFMove(cleanupFunction)));
+}
+
+ScreenCaptureSessionSource::ScreenCaptureSessionSource(WeakPtr<Observer>&& observer, RetainPtr<SCStream>&& stream, RetainPtr<SCContentFilter>&& filter, RetainPtr<SCContentSharingSession>&& sharingSession, CleanupFunction&& cleanupFunction)
+    : m_stream(WTFMove(stream))
+    , m_contentFilter(WTFMove(filter))
+    , m_sharingSession(WTFMove(sharingSession))
+    , m_observer(WTFMove(observer))
+    , m_cleanupFunction(WTFMove(cleanupFunction))
```

Comment:
It should probably be a CompletionHandler<> instead of a Function.

Reply:
Fixed